File: atcoder/abc176/e.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// 0-based index
// tree size: 2 * NN - 1
// leaves indices: [NN - 1, 2 * NN - 1)
template<class T>
struct SegTree {
    int NN;
    T default_value;
    vector<T> data;

    T aggr(T a, T b) {
        return max(a, b);
    }

    SegTree(int N, T val) {
        default_value = val;
        NN = 1; while (NN < N) NN <<= 1;
        data = vector<T>(2 * NN - 1, val);
    }

    SegTree(const vector<T> v, T val) {
        default_value = val;
        int N = v.size();
        NN = 1; while (NN < N) NN <<= 1;
        data = vector<T>(2 * NN - 1, val);
        copy(v.begin(), v.end(), data.begin() + (NN - 1));
        build(0, 0, NN);
    }

    void build(int u, int l, int r) {
        if (r - l == 1) return;
        int m = (l + r) / 2;
        build(u * 2 + 1, l, m);
        build(u * 2 + 2, m, r);
        data[u] = aggr(data[2 * u + 1], data[2 * u + 2]);
    }

    T query(int a, int b, int u, int l, int r) {
        if (l >= b || r <= a)
            return default_value;
        if (l >= a && r <= b)
            return data[u];
        int m = (l + r) / 2;
        T res1 = query(a, b, u * 2 + 1, l, m);
        T res2 = query(a, b, u * 2 + 2, m, r);
        return aggr(res1, res2);
    }

    void update(int idx, T x, int u, int l, int r) {
        if (idx < l || idx >= r)
            return;
        if (idx == l && r - l == 1) {
            data[u] = x;
            return;
        }
        int m = (l + r) / 2;
        update(idx, x, u * 2 + 1, l, m);
        update(idx, x, u * 2 + 2, m, r);
        data[u] = aggr(data[2 * u + 1], data[2 * u + 2]);
    }

    void increment(int idx, T x, int u, int l, int r) {
        update(idx, data[NN - 1 + idx] + x, u, l, r);
    }
};

template <class T>
ostream & operator << (ostream &out, const SegTree<T> &seg) {
    out << "SegTree(NN=" << seg.NN << ", ";
    for (int u = seg.NN - 1, sz = seg.NN; sz >= 1; u /= 2, sz /= 2) {
        out << "[ ";
        for (int i = 0; i < sz; i++) {
            out << seg.data[u + i] << " ";
        }
        out << "]";
    }
    out << ")";
    return out;
}

int solve() {
    int H, W, M;
    cin >> H >> W >> M;

    auto cntR = vector<int>(H, 0);
    auto cntC = vector<int>(W, 0);
    auto t = vector<vector<int>>(H, vector<int>());
    for (int i = 0; i < M; i++) {
        int r, c; 
        cin >> r >> c;
        r--; c--;
        cntR[r]++;
        cntC[c]++;
        t[r].push_back(c);
    }

    auto seg = SegTree<int>(cntC, 0);
    // cout << seg << endl;
    int ans = -1;
    for (int r = 0; r < H; r++) {
        for (auto c : t[r]) {
            seg.increment(c, -1, 0, 0, seg.NN);
        }
        // cout << seg << endl;
        int val = cntR[r] + seg.query(0, W, 0, 0, seg.NN);
        ans = max(ans, val);
        for (auto c : t[r]) {
            seg.increment(c, +1, 0, 0, seg.NN);
        }
    }
    return ans;
}

int main () {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cout << solve() << endl;

    return 0;
}

File: atcoder/abc176/e.py
class SegTree:
    def __init__(self, A, default, op):
        self.default = default
        self.op = op

        nn = 1
        while nn < len(A):
            nn *= 2
        self.nn = nn

        self.data = [default for _ in range(2 * nn)]
        self.data[nn - 1 : nn - 1 + len(A)] = A
        for u in range(nn - 2, -1, -1):
            self.data[u] = op(self.data[2 * u + 1], self.data[2 * u + 2])

    # query = [a, b), current at node u with [l, r)
    def query(self, a: int, b: int, u: int, l: int, r: int) -> int:
        # [l, r) not in [a, b)
        if r <= a or l >= b:
            return self.default
        # [l, r) is inside [a, b)
        if a <= l and r <= b:
            return self.data[u]
        # partially overlapped
        m = (l + r) // 2
        res1 = self.query(a, b, 2 * u + 1, l, m)
        res2 = self.query(a, b, 2 * u + 2, m, r)
        return self.op(res1, res2)

    # modify = [p, p + 1), current at node u with [l, r)
    def modify(self, p: int, x: int, u: int, l: int, r: int):
        # [l, r) not in [a, b)
        if r <= p or l >= p + 1:
            return
        # [l, r) is inside [a, b)
        if p <= l and r <= p + 1:
            self.data[u] += x
            return
        # partially overlapped
        m = (l + r) // 2
        self.modify(p, x, 2 * u + 1, l, m)
        self.modify(p, x, 2 * u + 2, m, r)
        self.data[u] = self.op(self.data[2 * u + 1], self.data[2 * u + 2])


H, W, M = map(int, input().split())
cntR = [0 for _ in range(H)]
cntC = [0 for _ in range(W)]
columns_in_row = [[] for _ in range(H)]
for _ in range(M):
    r, c = map(int, input().split())
    r, c = r - 1, c - 1
    cntR[r] += 1
    cntC[c] += 1
    columns_in_row[r].append(c)

# max_(0 <= r < H) max_(0 <= c < W) (cntR[r] + cntC[c] - A[r][c])
# = max_(0 <= r < H) (cntR[r] + max_(0 <= c < W) (cntC[c] - A[r][c]))

seg = SegTree(cntC, 0, max)
ans = 0
for r in range(H):
    # Subtract the duplicate count (A[r][c])
    for c in columns_in_row[r]:
        seg.modify(c, -1, 0, 0, seg.nn)

    val = cntR[r] + seg.query(0, W, 0, 0, seg.nn)
    ans = max(ans, val)

    # Restore the cntC
    for c in columns_in_row[r]:
        seg.modify(c, +1, 0, 0, seg.nn)

print(ans)
